File: CFSharp/CFSharp.F/Misc.fs
module CFSharp.F.Misc

open NUnit.Framework
open FsUnit
open System.Linq
open FSharpx.Strings

let mutable counter = 0

let IncCounter() = 
    counter <- counter + 1
    printfn "%d" counter

[<Test>]
let ``Look ma! Side-effects!``() =
    IncCounter()
    IncCounter()
    (* prints:
        1
        2       *)


let add a b = a + b

[<Test>]
let testAdd() = 
    add 1 2 |> should equal 3

let add2(a, b) = a + b

[<Test>]
let testAdd2() = 
    add2(1, 2) |> should equal 3

[<Test>]
let useLambda() =
    let list' = [1; 2; 3; 4; 5; 6; 7; 8; 9; 10]
    let list = [1..10]

    let result = list.Where(fun x -> x % 2 = 0)

    printfn "%A" result

let describe language =
    let description = 
        if language = "C#" then 
            "nice"
        else if language = "F#" then 
            "awesome"
        else if language = "Haskell" then 
            "begone, zealot!"
        else "I don't know that language"
    language + ": " + description

[<Test>]
let testDescribe() =
    describe "Haskell" |> should equal "Haskell: begone, zealot!"

[<Test>]
let loops() =
    let array = [|1; 2; 3; |]
    for x in array do
        // ...
        printfn "%A" x
    for x = 0 to array.Length-1 do
        // ...
        printfn "%d" array.[x]
    let mutable counter = 0
    while counter < 3 do
        // ...
        printfn "stuff!"
        counter <- counter+1

let topFsharpComments comments = 
    comments 
        |> Seq.filter (contains "F#") 
        |> Seq.truncate 5
        |> String.concat "\n"

[<Test>]
let testTopComments() =
    let result = topFsharpComments ([1..10] |> Seq.map (sprintf "blah blah F# %d"))
    printfn "%s" result

open System.IO

let foo fileName = 
    try
        use f = File.CreateText fileName
        f.WriteLine "hi!"
    with
        ex -> printfn "something went wrong: %A" ex

let commentsFor lang = [| "blah blah " + lang + " blah"; "something " + lang; "foo blah " + lang; |]
let commentsForLanguages languages =
    seq {
        for lang in languages do
        for comments in commentsFor lang do
        if contains "blah" comments then
            yield lang + ": " + comments
    }

[<Test>]
let testCommentsForLanguages() =
    let langs = [| "C#"; "F#" |]
    let result = String.concat "\n" (commentsForLanguages langs)
    printfn "%s" result
        
(* conditionals 
   loops
   assignments *)

(* sensible defaults:
        immutable
        no nulls
        *)

(* extra features:
        light-weight type alternatives
        pattern-matching
        ignored return (via ignore)
        *)

File: CFSharp/CFSharp.C/Misc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;
using Shouldly;
using System.IO;

namespace CFSharp.C
{
    public class Misc
    {
        private static int counter = 0;

        private static void IncCounter()
        {
            counter = counter + 1;
            Console.WriteLine("{0}", counter);
        }

        [Test]
        public void LookMaSideEffects()
        {
            IncCounter();
            IncCounter();
        }

        public static int Add(int a, int b)
        {
            return a + b;
        }

        [Test]
        public void TestAdd()
        {
            Add(1, 2).ShouldBe(3);
        }

        public static Func<int, int> CAdd(int a)
        {
            return b => a + b;
        }

        [Test]
        public void TestCurriedAdd()
        {
            CAdd(1)(2).ShouldBe(3);
        }

        private static int CAddOne(int b)
        {
            return CAdd(1)(b);
        }

        [Test]
        public void TestCurriedAddOne()
        {
            CAddOne(2).ShouldBe(3);
        }

        private static Func<int, Func<int, Func<int, int>>> Sum(int a)
        {
            return b => c => d => a + b + c + d;
        }
        
        [Test]
        public void TestSum()
        {
            Sum(1)(2)(3)(4).ShouldBe(10);
        }

        [Test]
        public void UseLambda()
        {
            var list = new[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
            var result = list.Where(x => x%2 == 0);

            Console.WriteLine(string.Join(", ", result));
        }

        private static string Describe(string language)
        {
            string description;
            if (language == "C#") {
                description = "nice";
            } else if (language == "F#") {
                description = "awesome";
            } else if (language == "Haskell") {
                description = "begone, zealot!";
            } else description = "I don't know that language"; 
            return language + ": " + description;
        }

        [Test]
        public void TestDescribe()
        {
            Describe("Haskell").ShouldBe("Haskell: begone, zealot!");
        }

        [Test]
        public void Loops()
        {
            var counter = 0;
            var array = new[] {1, 2, 3};
            foreach (var x in array) {
                // ...
                Console.WriteLine(x);
            }
            for (int i = 0; i < array.Length; i++) {
                // ...
                Console.WriteLine(array[i]);
            }
            while (counter < 3) {
                // ...
                Console.WriteLine("stuff");
                counter++;
            }
        }

        private static string TopFsharpComments(IEnumerable<string> comments)
        {
            var top = comments
                .Where(x => x.Contains("F#"))
                .Take(5);
            return string.Join("\n", top);
        }

        private static string TopFsharpComments2(IEnumerable<string> comments)
        {
            var filtered = comments .Where(x => x.Contains("F#"));
            var top = filtered.Take(5);
            return string.Join("\n", top);
        }

        private static IEnumerable<string> CommentsForLanguages(IEnumerable<string> languages)
        {
            return
                from lang in languages
                from comments in CommentsFor(lang)
                where comments.Contains("blah")
                select lang + ": " + comments;

        }

        private static IEnumerable<string> CommentsFor(string lang)
        {
            return new[] {
                "blah blah " + lang + " blah", 
                "something " + lang,
                "foo blah " + lang
            };
        }

        [Test]
        public void TestCommentsForLanguages()
        {
            Console.WriteLine(string.Join("\n", CommentsForLanguages(new[] {"C#", "F#"})));
        }

        private static void Foo(string fileName) {
            try {
                using (var f = File.CreateText(fileName)) {
                    f.WriteLine("hi!");
                }
            }
            catch (Exception ex) {
                Console.WriteLine("something went wrong: {0}", ex);
            }
        }

        [Test]
        public void ParseExample()
        {
            string s = "bob";
            int result;
            if (Int32.TryParse(s, out result))
            {
                //use result
            }
        }
    }
}